Extension templates can't include other files

Because the extension_report function uses render_template_string to render the extension template file contents, splitting an extension template into multiple files and using `{% include "sub_template.html" %}` isn't possible, since the extension's template directory isn't part of the Jinja/Flask search path. I tried a hack of temporarily setting fava_app.jinja_loader before calling render_template_string to inject the right search path:
```
current_loader = fava_app.jinja_loader
fava_app.jinja_loader = jinja2.ChoiceLoader(
    [
        jinja2.FileSystemLoader(
            [str(ext.extension_dir / "templates")]
        ),
        current_loader,
    ]
)
content = Markup(render_template_string(template, extension=ext))
fava_app.jinja_loader = current_loader
```
This works, and the render_template_string could probably also be switched to a render_template, but it isn't thread safe since I think it's possible for multiple requests get handled simultaneously in Flask. I haven't been able to find a good way to properly use more than one template directory within Flask. The closest thing I've found is that Blueprints can have their own template directories that get used in addition to the base template directory, so I think if a Blueprint got created and registered for each extension then this would be possible. The problem I can see with this is that the beancount file, and therefore the extensions, don't get loaded until the first request is made, at which point fava_app is obviously already running, so new Blueprints can't be registered. If the beancount file was loaded initially on startup then the Blueprints could be created and registered, and I think everything should work correctly. I can try to prepare a pull request that does all of this, but I wasn't sure if there was a good reason that the beancount file doesn't get loaded immediately that is more important than this.

I have a relatively straightforward change at tschicke/fava:c98f0d69210240b1aff0b73ba6758cd4dd36a44e that fixes this with a jinja2.FunctionLoader, but I don't see a way to get it to pass this linter error:
```src/fava/application.py:460: error: Incompatible types in assignment (expression has type "ChoiceLoader", variable has type "FileSystemLoader | None")  [assignment]```

--- a/src/fava/application.py
+++ b/src/fava/application.py
@@ -24,6 +24,7 @@ from urllib.parse import urlencode
 from urllib.parse import urlparse
 from urllib.parse import urlunparse
 
+import jinja2
 import markdown2  # type: ignore[import]
 from beancount import __version__ as beancount_version
 from beancount.utils.text_utils import replace_numbers
@@ -321,16 +322,16 @@ def _setup_routes(fava_app: Flask) -> None:  # noqa: PLR0915
     @fava_app.route("/<bfile>/extension/<extension_name>/")
     def extension_report(extension_name: str) -> str:
         """Endpoint for extension reports."""
-        ext = g.ledger.extensions.get_extension(extension_name)
-        if ext is None or ext.report_title is None:
+        g.extension = g.ledger.extensions.get_extension(extension_name)
+        if not g.extension or g.extension.report_title is None:
             abort(404)
-        template_path = ext.extension_dir / "templates" / f"{ext.name}.html"
-        template = template_path.read_text(encoding="utf-8")
-        content = Markup(render_template_string(template, extension=ext))
+        content = Markup(
+            render_template(f"{g.extension.name}.html", extension=g.extension),
+        )
         return render_template(
             "_layout.html",
             content=content,
-            page_title=ext.report_title,
+            page_title=g.extension.report_title,
         )
 
     @fava_app.route("/<bfile>/download-query/query_result.<result_format>")
@@ -420,6 +421,18 @@ def _setup_babel(fava_app: Flask) -> None:
         babel = Babel(fava_app, locale_selector=_get_locale)
 
 
+def _ext_template_loader_func(name: str) -> str | None:
+    """Load a template from the extension folder if an extension is active."""
+    if g.extension:
+        template_path = g.extension.extension_dir / "templates" / name
+        if template_path.exists():
+            return template_path.read_text(encoding="utf-8")
+    return None
+
+
+_ext_template_loader = jinja2.FunctionLoader(_ext_template_loader_func)
+
+
 def create_app(
     files: Iterable[Path | str],
     load: bool = False,
@@ -441,6 +454,12 @@ def create_app(
     _setup_babel(fava_app)
     _setup_filters(fava_app, read_only=read_only, incognito=incognito)
     _setup_routes(fava_app)
+    loader = fava_app.jinja_loader
+    if loader is None:
+        raise ValueError("Expected Flask app to have jinja_loader.")
+    fava_app.jinja_loader = jinja2.ChoiceLoader(  # type: ignore[assignment]
+        [loader, _ext_template_loader],
+    )
 
     fava_app.config["HAVE_EXCEL"] = HAVE_EXCEL
     fava_app.config["BEANCOUNT_FILES"] = [str(f) for f in files]

--- a/src/fava/context.py
+++ b/src/fava/context.py
@@ -8,17 +8,25 @@ from flask import g as flask_g
 if TYPE_CHECKING:  # pragma: no cover
     from fava.core import FavaLedger
     from fava.core import FilteredLedger
+    from fava.ext import FavaExtensionBase
     from fava.util.date import Interval
 
 
 class Context:
     """The allowed context values."""
 
+    #: Slug for the active Beancount file.
     beancount_file_slug: str | None
+    #: Conversion to apply
     conversion: str
+    #: Interval to group by
     interval: Interval
+    #: The ledger
     ledger: FavaLedger
+    #: The filtered ledger
     filtered: FilteredLedger
+    #: The current extension, if this is an extension endpoint
+    extension: FavaExtensionBase | None
 
 
 g: Context = flask_g  # type: ignore[assignment]

--- a/src/fava/ext/portfolio_list/templates/PortfolioList.html
+++ b/src/fava/ext/portfolio_list/templates/PortfolioList.html
@@ -1,6 +1,9 @@
 {% import "_query_table.html" as querytable %}
+
 {% set arg_filter = request.args.get('account_filter') %}
-<h2>Portfolio List Sample Report Extension</h2>
+
+{% include 'PortfolioListHeader.html' %}
+
 <br />
 <label for="portfolio-list-filter">Custom Account Filter:</label>
 <input id="portfolio-list-filter" value={{arg_filter or ""}}>

--- a/src/fava/ext/portfolio_list/templates/PortfolioListHeader.html
+++ b/src/fava/ext/portfolio_list/templates/PortfolioListHeader.html
@@ -1,0 +1,1 @@
+<h2>Portfolio List Sample Report Extension</h2>

